blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,866 @@
+            local_addrs: Vec::new(),
+            pending: HashMap::new(),
+            cur_tipset: tipset,
+            api,
+            min_gas_price: ToBigInt::to_bigint(&0).unwrap(),
+            max_tx_pool_size: 5000,
+            network_name,
+            bls_sig_cache,
+            sig_val_cache,
+            local_msgs: HashMap::new(),
+        };
+        mp.load_local()?;
+        Ok(mp)
+    }
+
+    /// Add a signed message to local_addrs and local_msgs
+    fn add_local(&mut self, m: &SignedMessage, msgb: Vec<u8>) -> Result<(), Error> {
+        self.local_addrs.push(*m.from());
+        self.local_msgs.insert(msgb, m.clone());
+        Ok(())
+    }
+
+    /// Push a signed message to the MessagePool
+    pub fn push(&mut self, msg: &SignedMessage) -> Result<Cid, Error> {
+        let msg_serial = msg
+            .marshal_cbor()
+            .map_err(|err| Error::Other(err.to_string()))?;
+        self.add(msg)?;
+        self.add_local(msg, msg_serial)?;
+        msg.cid().map_err(|err| Error::Other(err.to_string()))
+    }
+
+    /// This is a helper to push that will help to make sure that the message fits the parameters
+    /// to be pushed to the MessagePool
+    pub fn add(&mut self, msg: &SignedMessage) -> Result<(), Error> {
+        let size = msg
+            .marshal_cbor()
+            .map_err(|err| Error::Other(err.to_string()))?
+            .len();
+        if size > 32 * 1024 {
+            return Err(Error::MessageTooBig);
+        }
+        if msg
+            .value()
+            .gt(&ToBigUint::to_biguint(&2_000_000_000).unwrap())
+        {
+            return Err(Error::MessageValueTooHigh);
+        }
+
+        self.verify_msg_sig(msg)?;
+
+        let tmp = msg.clone();
+        self.add_tipset(tmp, &self.cur_tipset.clone())
+    }
+
+    /// Add a SignedMessage without doing any of the checks
+    pub fn add_skip_checks(&mut self, m: SignedMessage) -> Result<(), Error> {
+        self.add_helper(m)
+    }
+
+    /// Return the string representation of the message signature
+    pub(crate) fn sig_cache_key(&mut self, msg: &SignedMessage) -> Result<String, Error> {
+        match msg.signature().signature_type() {
+            SignatureType::Secp256k1 => Ok(msg.cid().unwrap().to_string()),
+            SignatureType::BLS => {
+                if msg.signature().bytes().len() < 90 {
+                    return Err(Error::BLSSigTooShort);
+                }
+                let mut beginning = String::new();
+                beginning += &msg
+                    .cid()
+                    .unwrap()
+                    .to_bytes()
+                    .into_iter()
+                    .map(char::from)
+                    .collect::<String>();
+                beginning += &msg
+                    .signature()
+                    .bytes()
+                    .iter()
+                    .map(|b| char::from(*b))
+                    .collect::<String>();
+                Ok(beginning)
```

Comment:
once previous changes, you should just be appending the signature bytes to the cid bytes